lte/gateway/c/core/oai/tasks/async_grpc_service/grpc_async_service_task.cpp

```diff
@@ -0,0 +1,232 @@
+    : cq_(std::move(cq)) {}
+void AsyncService::wait_for_requests() {
+  init_call_data();
+  void* tag;
+  bool ok;
+  running_ = true;
+  while (running_) {
+    if (cq_ == nullptr) {
+      OAILOG_CRITICAL(LOG_S6A, "Completion queue on s6a interface is null \n");
+      return;
+    }
+    if (!(cq_->Next(&tag, &ok))) {
+      OAILOG_ERROR(
+          LOG_S6A, "Completion queue on s6a interface is shutting down \n");
+      return;
+    }
+    if (!ok) {
+      continue;
+    }
+    static_cast<CallData*>(tag)->proceed();
+  }
+}
+
+void AsyncService::stop() {
+  running_ = false;
+  // Pop all items in the queue until it is empty
+  void* tag;
+  bool ok;
+  while (cq_->Next(&tag, &ok)) {
+  }
+}
+S6aProxyResponderAsyncService::S6aProxyResponderAsyncService(
+    std::unique_ptr<ServerCompletionQueue> cq,
+    std::shared_ptr<S6aProxyAsyncResponderHandler> handler)
+    : AsyncService(std::move(cq)), handler_(handler) {}
+void S6aProxyResponderAsyncService::init_call_data(void) {
+  new CancelLocationCallData(cq_.get(), *this, *handler_);
+  new ResetCallData(cq_.get(), *this, *handler_);
+}
+
+void S6aProxyResponderAsyncService::set_callback(
+    std::unique_ptr<ServerCompletionQueue> cq,
+    std::shared_ptr<S6aProxyAsyncResponderHandler> handler) {
+  this->cq_      = std::move(cq);
+  this->handler_ = handler;
+}
+
+template<class GRPCService, class RequestType, class ResponseType>
+AsyncGRPCRequest<GRPCService, RequestType, ResponseType>::AsyncGRPCRequest(
+    ServerCompletionQueue* cq, GRPCService& service)
+    : cq_(cq), status_(PROCESS), responder_(&ctx_), service_(service) {}
+
+// Internal state management logic for AsyncGRPCRequest
+// Once a request has started processing, create a new AsyncGRPCRequest to
+// standby for new requests. After a request has finished processing, delete the
+// object.
+template<class GRPCService, class RequestType, class ResponseType>
+void AsyncGRPCRequest<GRPCService, RequestType, ResponseType>::proceed() {
+  if (status_ == PROCESS) {
+    clone();  // Create another stand by CallData
+    process();
+    status_ = FINISH;
+  } else {
+    GPR_ASSERT(status_ == FINISH);
+    delete this;
+  }
+}
+
+template<class GRPCService, class RequestType, class ResponseType>
+std::function<void(grpc::Status, ResponseType)> AsyncGRPCRequest<
+    GRPCService, RequestType, ResponseType>::get_finish_callback() {
+  return [this](grpc::Status status, ResponseType response) {
+    responder_.Finish(response, status, (void*) this);
+  };
+}
+
+void S6aProxyAsyncResponderHandler::CancelLocation(
+    ServerContext* context, const CancelLocationRequest* request,
+    std::function<void(grpc::Status, CancelLocationAnswer)> response_callback) {
+  auto& request_cpy = *request;
+  CancelLocationAnswer ans;
+  Status status;
+  status = Status::OK;
+  ans.set_error_code(ErrorCode::SUCCESS);
+  if (response_callback) {
+    response_callback(status, ans);
+  }
```

Comment:
if we are responding back like this immediately, what is the purpose of having an async call?

Reply:
Hey @ulaskozat  note the issue is `send_cancel_location_ans` not implementing the cancel location answer. 

Per my understanding  we needed an async service to be able to answer back with the proper error code. @rsarwad 
 can you confirm?

http://github.com/magma/magma/blob/master/lte/gateway/c/core/oai/tasks/s6a/s6a_grpc_iface.cpp#L48-L48

Reply:
Yes @uri200, what you say is right. Earlier AGW was not responding to cancel location request. As part of this PR, we added code to send Cancel location answer in response to cancel location request, with result code "DIAMETER_SUCCESS" because as per spec, https://www.etsi.org/deliver/etsi_ts/129200_129299/129272/12.06.00_60/ts_129272v120600p.pdf section, 5.2.1.2.2 
It indicates that MME always send Cancel location answer with result code "DIAMETER_SUCCESS".
So we can send the response immediately